text/0000-full-wallet-lifecycle.md

```diff
@@ -0,0 +1,195 @@
+~/.grin/main/
+   grin-wallet.log
+   grin-wallet.toml
+   wallet_data/
+     wallet.seed
+     db/
+     saved_txs/
+```
+
+`wallet_data` contains the seed and database for a single instance of the wallet, with the path to the wallet's data directory configured in
+`grin_wallet.toml`.
+
+The structure of the wallet's data files will be changed to allow multiple wallet instances, with each named wallet contained in a directory
+named according to name of the wallet provided by the end user. Current single-instance wallet data will be placed into a directory called
+`default`, and the API and all wallet software will assume the default directory is to be used if no name is provided.
+
+The wallet data directory structure will become (for example):
+
+```
+~/.grin/main/
+   grin-wallet.log
+   grin-wallet.toml
+   wallet_data/
+     default/
+        wallet.seed
+        db/
+        saved_txs/
+     my_wallet_1/
+        wallet.seed
+        ...
+     my_wallet_2/
+        wallet.seed
+        ...
+```
+
+If a data directory is not provided in a wallet API call, it will be assumed to be operating on the `default` wallet.
+
+There should only be a single `grin-wallet.toml` file in the wallet's data directory. Its configuration settings are global to all wallet
+instances.
+
+### Data migration
+
+The target version of the wallet will contain a function to migrate existing wallets from the current data structure to the new,
+essentially moving `wallet_data/wallet.seed`, `wallet_data/db/` and `wallet_data/saved_txs/` into `wallet_data/default/wallet.seed` etc.
+(Should this be performed within an API call for consistency?)
+
+## Wallet Initialization
+
+Currently, wallet data does not exist until the user runs `grin-wallet init`. The `init` command creates `grin-wallet.toml`,
+in the `~/.grin/main` directory (or `~/.grin/floonet`, or the current directory via the `-h` flag), prompts the user for a password,
+creates a seed file, stores the resulting data files in the directory specified in `grin-wallet.toml` (`~/.grin/main/wallet_data` by default)
+and initialises the lmdb database.
+
+It should be possible to run `grin-wallet owner_api` or invoke the API directly from a linked binary without having instantiated a wallet.
+
+(How is grin-wallet.toml generated?)
+
+## Wallet Runtime Instantiation
+
+Currently, wallet instantiation works differently depending on whether the wallet is invoked via the APIs or via the command line.
+
+1. In the command line case, each wallet command is a separate invocation. Command line invocation will ask for a password, decrypt the master seed and initialize the wallet with the descrypted seed. It will then perform the desired function and return, zeroing memory and exiting the process.
+
+1. In the case of a listening API (owner or foreign), the password is given once and the wallet seed is decrypted. The wallet instance is then kept in memory by the handling thread, and re-used for each Foreign or Owner API call.
```

Comment:
Should probably be a 2 here?

Reply:
Nope, this is [markdown syntax](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#lists). :) It renders well in the [actual document](https://github.com/mimblewimble/grin-rfcs/blob/7a7b2aae944edcbd37b7f117724ca92d9d151e76/text/0000-full-wallet-lifecycle.md#wallet-runtime-instantiation).

Reply:
Oh it totally does. 😄 Didn't know that markdown fixed list indexes like that.